products/apigateway/api.yaml

```diff
@@ -0,0 +1,227 @@
+      - !ruby/object:Api::Type::Time
+        name: 'createTime'
+        description: Creation timestamp in RFC3339 text format.
+        output: true
+      - !ruby/object:Api::Type::KeyValuePairs
+        name: 'labels'
+        description: |
+          Resource labels to represent user-provided metadata.
+  - !ruby/object:Api::Resource
+    name: 'ApiConfig'
+    min_version: beta
+    create_url: projects/{{project}}/locations/global/apis/{{api}}/configs?apiConfigId={{api_config_id}}
+    self_link: projects/{{project}}/locations/global/apis/{{api}}/configs/{{api_config_id}}
+    base_url: projects/{{project}}/locations/global/apis/{{api}}/configs
+    read_query_params: '?view=FULL'
+    update_verb: :PATCH
+    update_mask: true
+    description: |
+      A consumable API that can be used by multiple Gateways.
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      guides:
+        'Official Documentation':
+          'https://cloud.google.com/api-gateway/docs/quickstart'
+      api: 'https://cloud.google.com/api-gateway/docs/reference/rest/v1beta/projects.locations.apis.configs'
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: api
+        description: |
+          The API to attach the config to.
+        required: true
+        input: true
+        url_param_only: true
+      - !ruby/object:Api::Type::String
+        name: apiConfigId
+        description: |
+          Identifier to assign to the API Config. Must be unique within scope of the parent resource.
+        required: true
+        input: true
+        url_param_only: true
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: The resource name of the API Config.
+        output: true
+      - !ruby/object:Api::Type::String
+        name: 'displayName'
+        description: |
+          A user-visible name for the API.
+      - !ruby/object:Api::Type::String
+        name: 'serviceConfigId'
+        output: true
+        description: |
+          The ID of the associated Service Config (https://cloud.google.com/service-infrastructure/docs/glossary#config).
+      - !ruby/object:Api::Type::KeyValuePairs
+        name: 'labels'
+        description: |
+          Resource labels to represent user-provided metadata.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'gatewayConfig'
+        description: |
+          Immutable. Gateway specific configuration.
+          If not specified, backend authentication will be set to use OIDC authentication using the default compute service account
+        properties:
+        - !ruby/object:Api::Type::NestedObject
+          name: 'backendConfig'
+          description: |
+            Backend settings that are applied to all backends of the Gateway.
+          properties:
+          - !ruby/object:Api::Type::String
+            name: 'googleServiceAccount'
```

Comment:
same here

Reply:
No, it defaults to the GCE default service account.

Same with the parent field.

Reply:
That true if the user doesn't specify a `gatewayConfig` though, right? If so, then we can just allow users that don't set one to have the API-given default, and require users that do set it to actually specify the field.

Reply:
i see, make it required on googleServiceAccount but not on backendConfig

Reply:
I guess my question is, is there a functional difference between these three scenarios?
1:
```
gatewayConfig: null
```

2:
```
gatewayConfig: {}
```

3:
```
gatewayConfig: {
  backendConfig: {}
}
```

Reply:
The null is tested in the basic example and there are no problems there.

The other two.

```
 REDACTED  MCW0CDP3YY  ~  $  echo '{"openapiDocuments":[{"document":{"contents":"c3dhZ2dlcjogJzIuMCcKaW5mbzoKICB0aXRsZTogZGV2IAogIGRlc2NyaXB0aW9uOiBTYW1wbGUgQVBJIG9uIEFQSSBHYXRld2F5IHdpdGggYSBDbG91ZCBSdW4gYmFja2VuZAogIHZlcnNpb246IDEuMC4wCnNjaGVtZXM6CiAgLSBodHRwcwpwcm9kdWNlczoKICAtIGFwcGxpY2F0aW9uL2pzb24KcGF0aHM6CiAgL2hlbGxvOgogICAgZ2V0OgogICAgICBzdW1tYXJ5OiBHcmVldCBhIHVzZXIKICAgICAgb3BlcmF0aW9uSWQ6IGhlbGxvCiAgICAgIHgtZ29vZ2xlLWJhY2tlbmQ6CiAgICAgICAgYWRkcmVzczogaHR0cHM6Ly9nb29nbGUuY29tCiAgICAgIHJlc3BvbnNlczoKICAgICAgICAnMjAwJzoKICAgICAgICAgIGRlc2NyaXB0aW9uOiBBIHN1Y2Nlc3NmdWwgcmVzcG9uc2UKICAgICAgICAgIHNjaGVtYToKICAgICAgICAgICAgdHlwZTogc3RyaW5n","path":"spec.yaml"}}],"gatewayConfig":{"backendConfig":{}}}' | http POST "https://apigateway.googleapis.com/v1beta/projects/REDACTED/locations/global/apis/test/configs?alt=json&apiConfigId=potato" "Authorization: Bearer $(gcloud auth print-access-token)"
HTTP/1.1 200 OK
Alt-Svc: h3-Q050=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-27=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
Cache-Control: private
Content-Encoding: gzip
Content-Type: application/json; charset=UTF-8
Date: Wed, 21 Oct 2020 18:35:45 GMT
Server: ESF
Transfer-Encoding: chunked
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 0

{
    "done": false,
    "metadata": {
        "@type": "type.googleapis.com/google.cloud.apigateway.v1beta.OperationMetadata",
        "apiVersion": "v1beta",
        "createTime": "2020-10-21T18:35:45.486637683Z",
        "requestedCancellation": false,
        "target": "projects/REDACTED/locations/global/apis/test/configs/potato",
        "verb": "create"
    },
    "name": "projects/REDACTED/locations/global/operations/operation-1603305344956-5b2329d8d7825-80f08ca0-cb70a3e5"
}

 REDACTED  MCW0CDP3YY  ~  SIGINT  $  echo '{"openapiDocuments":[{"document":{"contents":"c3dhZ2dlcjogJzIuMCcKaW5mbzoKICB0aXRsZTogZGV2IAogIGRlc2NyaXB0aW9uOiBTYW1wbGUgQVBJIG9uIEFQSSBHYXRld2F5IHdpdGggYSBDbG91ZCBSdW4gYmFja2VuZAogIHZlcnNpb246IDEuMC4wCnNjaGVtZXM6CiAgLSBodHRwcwpwcm9kdWNlczoKICAtIGFwcGxpY2F0aW9uL2pzb24KcGF0aHM6CiAgL2hlbGxvOgogICAgZ2V0OgogICAgICBzdW1tYXJ5OiBHcmVldCBhIHVzZXIKICAgICAgb3BlcmF0aW9uSWQ6IGhlbGxvCiAgICAgIHgtZ29vZ2xlLWJhY2tlbmQ6CiAgICAgICAgYWRkcmVzczogaHR0cHM6Ly9nb29nbGUuY29tCiAgICAgIHJlc3BvbnNlczoKICAgICAgICAnMjAwJzoKICAgICAgICAgIGRlc2NyaXB0aW9uOiBBIHN1Y2Nlc3NmdWwgcmVzcG9uc2UKICAgICAgICAgIHNjaGVtYToKICAgICAgICAgICAgdHlwZTogc3RyaW5n","path":"spec.yaml"}}],"gatewayConfig":{}}' | http POST "https://apigateway.googleapis.com/v1beta/projects/REDACTED/locations/global/apis/test/configs?alt=json&apiConfigId=tf-test-api-cfgvjqjfh298j" "Authorization: Bearer $(gcloud auth print-access-token)"
HTTP/1.1 200 OK
Alt-Svc: h3-Q050=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-27=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
Cache-Control: private
Content-Encoding: gzip
Content-Type: application/json; charset=UTF-8
Date: Wed, 21 Oct 2020 18:31:53 GMT
Server: ESF
Transfer-Encoding: chunked
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 0

{
    "done": false,
    "metadata": {
        "@type": "type.googleapis.com/google.cloud.apigateway.v1beta.OperationMetadata",
        "apiVersion": "v1beta",
        "createTime": "2020-10-21T18:31:53.702041935Z",
        "requestedCancellation": false,
        "target": "projects/REDACTED/locations/global/apis/test/configs/tf-test-api-cfgvjqjfh298j",
        "verb": "create"
    },
    "name": "projects/REDACTED/locations/global/operations/operation-1603305113163-5b2328fbc959d-c44b83b5-1c27fca5"
}

```

Reply:
It's not about whether or not there are problems, but about providing a consistent interface to our users and helping them write their configs correctly. If there are two ways of doing the exact same thing, we can simplify and make it so that there's only one way. https://github.com/hashicorp/terraform-provider-google/issues/3928 has more context on why we try not to allow objects with only optional fields. It's a thing we allow only when there is a specific need.

Reply:
I don't follow.

Are you saying make all the child fields mandatory and the top level field (gatewayConfig) optional?

Reply:
Yup!

Reply:
Cool, did that in my last commit.